### DISCUSSION TITLE
A case for xref macros

### DISCUSSION THREAD

**Discussion by Josh-Cena** [mdn/mdn-community#350] [Platform]
This is a case of why I hope xref macros would not be removed, and even be encouraged of using. We've recently seen a lot of discussions on why people don't like xrefs and how we want to deprecate them, but IMO their benefits greatly outweigh the complexity.

## On moving to Markdown links

There are three reasons why xrefs are better than Markdown links.

### Shorter source

Shorter Markdown source has multiple benefits:

- Better gauging of content length. Sometimes, writers create paragraphs that are too long or too short, and this severely impedes the flow/rhythm of the prose. The Markdown source's length is a quick way to estimate how long the paragraphs will be, and ensure that they are approximately balanced. However, if a paragraph contains several links, the source length grows much more quickly. This is especially the case for JS docs, where each page has a looong slug prefix:

```md
[`Array`](/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array)
{{jsxref("Array")}}
Array
```

- Tables. Using tables as a list for API pages is a very plausible use-case, especially when there are more than two columns, which means we can't use a DL. However, the Markdown style guidelines require Markdown tables to have a maximum line length of 150 characters. The single link above is already half of that (69 characters), and it renders to merely 5 characters! This makes Markdown links and tables fundamentally incompatible.
- Unnecessary duplication. The source is not only long, it's also unnecessarily duplicated. Duplication creates room for errors. Consider some subtle bugs:

```md
[`Object.keys()`](/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/values)
```

Such errors may result from careless copy/pasting. However, the fundamental issue here is that the same piece of information—a link to the `Object.keys()` page—is repeated twice, once in the title, once in the URL. If we replace it with `{{jsxref("Object.keys()")}}`, then there's no duplication altogether, and thus no room for typos.

### Decoupling from content structure

Whether slug-based or file-location-based, Markdown links inevitably involve typing out the entire path to the target page. This means any relocation of a page requires updating every link to it, which is annoying.

```diff
- [`Array`](/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array)
+ [`Array`](/en-US/docs/Web/JavaScript/Reference/Globals/Array)
  {{jsxref("Array")}} <!-- No change! -->
```

I acknowledge that the current xref implementation is far from ideal. @wbamberg once mentioned that the CSS xref forced the docs to have a flat structure. However, such implementation details should not mask away the underpinning idea of an xref: to abstract away the file system structure. Take a CSS xref for example: `{{cssxref("accent-color")}}`. Nothing of its API surface requires a particular file structure, and we can simply re-implement the xref so it does a search among various potential locations for the `accent-color` page. Fix issues of xrefs by fixing them—not by staying away from them 🙂

### Ease of authoring

JS links are long, and to this day I don't remember the prefix syntax. Is it `/en-US/docs/Web/`? Is it `/en-US/Web/docs/`? Is it `/Reference/` or `/References/`? In the end, the most effective way is to go to the respective page on the website and copy its URL, but that involves switching windows and navigating across pages. On the other hand, using xrefs, all I need is to type out exactly what's on my mind: a link to `Array`.

@onkarruikar and @teoli2003 advocated for file-location-based links because of the better editor support. However, this still requires tabbing through every path segment: with a path like `files/en-us/web/javascript/reference/global_objects/array/index.md`, that would be 8 tabs, plus some characters to reduce the search space for autocompletion. OTOH, because VS Code also auto-closes braces and parentheses, `{{jsxref("Array")}}` itself only takes 15 keystrokes, which can be further reduced if one has an editor snippet for `{{jsxref("")}}`.

The idea is to increase information entropy as much as possible. Make every character typed meaningful!

## On creating a better xref macro

I can only speak for the jsxref macro in terms of implementation, because I haven't used others as much, but I believe many ideas also apply. There are a few common patterns that lead to the macro not being as ergonomic as it can be:

1. `/`-separated segments should never be present in the output. Sometimes, a prefix is necessary to disambiguate—for example, `{{jsxref("Operators/in")}}` instead of `{{jsxref("in")}}`, and `{{jsxref("Array/map")}}` instead of `{{jsxref("map")}}`. However, this means the rendered text is also `Operators/in`, which makes no sense. A smart xref should recognize that the `Operators/` part is only present for namespacing, and only render `in`, so we don't have to write `{{jsxref("Operators/in", "in")}}`.
2. Understanding trailing `()`. Constructors and classes have separate pages—for example, `{{jsxref("Array/Array")}}` and `{{jsxref("Array")}}`. In prose, to differentiate them at a glance, we always stylize the former as `Array()` (with the `()`). We can take advantage of this and make `{{jsxref("Array()")}}` automatically expand to `{{jsxref("Array/Array")}}`. This gets rid of the `Array/Array` duplication.
3. Underscores. Sometimes, writers write a link like `{{jsxref("Statements/async_function")}}` and unintentionally introduce the underscores to the output. However, as a matter of fact, we know that the slug can never have spaces, so `{{jsxref("Statements/async function")}}` should just as unambiguously point to the same page, while not requiring repeating the title a second time for the rendered text argument.
4. No-code format. Right now, the switch is the last argument, which means if we want no-code, we must use all previous three arguments (target, rendered text, and anchor), even when they are empty or repeated. This can be fixed by requiring the no-code flag to be a number/boolean, so whenever we encounter such an input, we know it's the last. This allows us to refactor `{{jsxref("Inheritance and the prototype chain", "Inheritance and the prototype chain", "", 1)}}` to `{{jsxref("Inheritance and the prototype chain", 1)}}`.

> **Note** I would gladly send a PR to Yari adding the said features if people find them promising.

This makes the xref much more complex, you may say. That's true, but these are complexities that authors will enjoy. What confuses authors the most about xrefs is not their behavior, but the poor documentation. If we can have a dedicated page on the various features of xrefs, I believe authors will soon start to enjoy the awesome experience xrefs bring about.

**Comment by OnkarRuikar** [comment]
> Better gauging of content length...

Almost every IDE has preview feature, including GitHub. Why not rely on that?
Without the macros, authors will be able to see exactly how the page would render. They won't have to wait for PR companion or run local yari(`yarn start`), so it will be quicker.
![bug](https://user-images.githubusercontent.com/87750369/226501183-2069146d-7748-4072-a2ae-2203cd88a57e.png)

Also, macros are counter productive in such previews as they are not rendered:
![bug](https://user-images.githubusercontent.com/87750369/226512455-24a56d3d-82dd-473e-a9e2-78dde865ee4e.png)

***

> Tables. ... However, the Markdown style guidelines require Markdown tables to have a maximum line length of 150 characters. The single link above is already half of that (69 characters), and it renders to merely 5 characters!

At the moment the markdown linter doesn't enforce any line lengths. `MD013 - Line length` rule is disabled in config. In case we enable it line lengths for tables can be disabled or set.

***

> Unnecessary duplication. The source is not only long, it's also unnecessarily duplicated. Duplication creates room for errors.

To avoid duplication reference style links can be used:
```md
## Get object keys

Function [`Object.keys()`][object.keys] can be used to get all object keys.\
[`keys()`][1] function was added in ES version xyz.\
You can override [`Object.keys()`][1].

## More sections below

...

[1]: /files/en-us/web/javascript/reference/global_objects/keys
[object.keys]: /files/en-us/web/javascript/reference/global_objects/keys
```
This will be way shorter than macro calls. Also, it'll reduce errors due to copy paste as you need to get only one link correct. Using default language features is more efficient.

***

> **Decoupling from content structure**\
>  This means any relocation of a page requires updating every link to it, which is annoying.

- This is being addressed in yari `yarn content move` command: https://github.com/mdn/yari/issues/8292. Once this https://github.com/mdn/yari/pull/8412 gets in I'll work on move as the changes are similar.\
It's easy to update markdown style links than updating macros when a document is moved to completely new location.

***
> **Ease of authoring**\
> JS links are long, and to this day I don't remember the prefix syntax.\
> @OnkarRuikar and @teoli2003 advocated for file-location-based links because of the better editor support. However, this still requires tabbing through every path segment

I feel the pain. Again reference style links(explained above) can be used. You need to get the link right only once.\
For macro lovers we can create emmet snippets:

https://user-images.githubusercontent.com/87750369/226529494-2b0c80ad-35eb-46fb-a34c-36c596e701cb.mp4

*Ignore html comment style expansions. It is work in progress.\
Notice the tab stops first one in link text and second one at the end of URL part.

***
> This makes the xref much more complex, you may say. That's true, but these are complexities that authors will enjoy. 

New/non-frequent contributors will still have to refer the macro guide page and go through file hierarchy to craft working macro call. They'll still have to verify it using `yarn start` or PR companion.

While working on removing html xref macro we noticed that people guessed page locations in macro calls and didn't bother to verify link validity later. \
If we stick to natural markdown style syntax, contributors need to look at only the file hierarchy. The `xrefs` are kinda looking like rocket science as every technology/language has separate macro, params, and file hierarchy.

***
## Case for automation

1. Macros are hard to validate, the only way to do it is using yari flaws tool. If we use markdown style links then IDEs do point out broken links while editing itself.
2. Daily/weekly auto fix automation is possible if we stick to markdown syntax.

**Comment by wbamberg** [comment]
> I acknowledge that the current xref implementation is far from ideal. @wbamberg once mentioned that the CSS xref forced the docs to have a flat structure. However, such implementation details should not mask away the underpinning idea of an xref: to abstract away the file system structure. Take a CSS xref for example: {{cssxref("accent-color")}}. Nothing of its API surface requires a particular file structure, and we can simply re-implement the xref so it does a search among various potential locations for the accent-color page. Fix issues of xrefs by fixing them—not by staying away from them 🙂

That's fair. At the moment xrefs are tied to file/URL structure, but yes, that's because we don't have an organized way of saying what kinds of items are supposed to be xreffable, so our usage rules collapse into their current implementation.

In reality I suppose any group of items which share a namespace could be xreffable. This would mean we'd have to:
- design clear rules about what xrefs can be used for
- have a way to deal with items in related but distinct namespaces
- be prepared to accept a potentially much more complicated implementation, with a properly abstracted API

For instance, if we wanted `cssxref` to be usable for properties ([`color`](https://developer.mozilla.org/en-US/docs/Web/CSS/color)) and types ([`<color>`](https://developer.mozilla.org/en-US/docs/Web/CSS/color_value)), we'd need a way to accommodate that. A bit harder would be for static and non-static methods, like `Response.json()` (which we can't document now, embarrassingly, but obviously need to).

**Reply by wbamberg** [reply]
This is quite half-baked, but maybe we could start by saying:
- there's some finite set of sets of things that are xreffable (CSS properties, HTML elements, Web/API interfaces)
- each of those things has a namespace inside which things have unique names
- an xref (maybe just one xref, rather than one for each domain) then needs to be given the namespace and the name within that namespace (like "CSS properties", "margin"), and that's enough for the xref to generate a link
- maybe page types could help here, because they generally identify namespaces?

Yari could for instance get the set of `web-api-interface` pages, and when it sees `{{xref("web-api-interface", "Request")}}` could find the page that references.

There are places where page types aren't that helpful: for instance we have different page types for `css-shorthand-property` although they share a namespace with `css-property`. Maybe we would want to revise page types in these cases?

**Reply by OnkarRuikar** [reply]
> Yari could for instance get the set of `web-api-interface` pages, and when it sees `{{xref("web-api-interface", "Request")}}` could find the page that references.

This sounds nice as it gives us the freedom of creating any directory hierarchy.\
But this isn't foolproof as there are multiple elements with same name for same page-types:
```
slug: Web/API/FetchEvent/request
page-type: web-api-instance-property

slug: Web/API/IDBCursor/request
page-type: web-api-instance-property
```
```
slug: Web/API/LockManager/request
page-type: web-api-instance-method

slug: Web/API/WakeLock/request
page-type: web-api-instance-method
```
If we use it like `{{xref("web-api-instance-property", "APIname.prop")}}` then it could work.\
But then the second arg format will have to change depending on the page type e.g. `{{xref("web-api-interface", "APIname")}}` vs `{{xref("web-api-method", "APIname.method")}}`. So it still remains complex, as you have to deal with multiple page types compared to only cssxref, jsxref etc.

***

There are 575 xref macro errors at the moment in `mdn/content`. Are we fixing those while implementing new `xref` mechanism?

No one tries to fix flaws mentioned in PR companion. Having simpler mechanisms is beneficial in the long term. Pure markdown syntax is always easy to maintain.

**Reply by wbamberg** [reply]
> But then the second arg format will have to change depending on the page type e.g. {{xref("web-api-interface", "APIname")}} vs {{xref("web-api-method", "APIname.method")}}.

But this is already the case, implicitly, with domxref (`{{domxref("Request")}}` versus `{{domxref("Request.body")}}`). So yes, page type doesn't exactly identify a namespace in situations like this. But you could I think still use it, to distinguish for example instance and static methods, which can have the same name in an interface.

Historically I've been team Markdown links, but I think Josh makes good points, and to some extent his post has made me realise that what I really don't like isn't xrefs so much as the random way they are implemented and used in MDN.

**Reply by OnkarRuikar** [reply]
> Historically I've been team Markdown links, but I think Josh makes good points, and to some extent his post has made me realise...

Oh noo, don't switch the team yet! Is it because the ease of use or how the content looks(pretty) using xrefs?

Just a thought. How about we, for ease of use, have contributors keep using xref macros but at end of the day or week we run a script to convert `{{xref("AbcAPI")}}` to ```[`AbcAPI`](/en-US/../abcapi)```.
